vlib/os/process.v

```
@@ -0,0 +1,240 @@
+module os
+
+// ProcessState.not_started - the process has not yet started
+// ProcessState.running - the process is currently running
+// ProcessState.stopped - the process was running, but was stopped temporarily
+// ProcessState.exited - the process has finished/exited
+// ProcessState.aborted - the process was terminated by a signal
+pub enum ProcessState {
+	not_started
+	running
+	stopped
+	exited
+	aborted
+}
+
+[ref_only]
+pub struct Process {
+pub:
+	filename      string // the process's command file path
+pub mut:
+	pid           int // the PID of the process
+	code          int = -1
+	// the exit code of the process, != -1 *only* when status is .exited *and* the process was not aborted
+	status        ProcessState = .not_started
+	// the current status of the process
+	err           string // if the process fails, contains the reason why
+	args          []string // the arguments that the command takes
+	env_is_custom bool // true, when the environment was customized with .set_environment
+	env           []string // the environment with which the process was started
+	use_stdio_ctl bool // when true, then you can use p.stdin_write(), p.stdout_slurp() and p.stderr_slurp()
+	stdio_fd      [3]int
+}
+
+// new_process - create a new process descriptor
+// NB: new does NOT start the new process.
+// That is done because you may want to customize it first,
+// by calling different set_ methods on it.
+// In order to start it, call p.run() or p.wait()
+pub fn new_process(filename string) &Process {
+	return &Process{
+		filename: filename
+	}
+}
+
+// set_args - set the arguments for the new process
+pub fn (mut p Process) set_args(pargs []string) &Process {
+	if p.status != .not_started {
+		return p
```

Isn't it better to return an error in this case and similar cases? It's more explicit and easier to track possible bugs for devs if we return an error instead of ignoring the whole method call in case of being in any state other than not started.

P.S: I'm new to V and this community and I may be wrong. I appreciate if you explain this to me if I'm wrong or it's against conventions or I'm missing something or there are drawbacks doing so.

---

The initial version was from 6 months ago, and then V allowed chaining of methods called on mut variables.
Using this, the API was written, so you could do:
`os.new_process('/bin/sleep').set_args(['20']).run()`

V currently does not allow that anymore, and returning `?` may be better and more robust indeed. I intend to clean up the API after windows is supported too.

---

Chaining like that will be allowed again, but it's not really the recommended V way. We don't have APIs designed this way, so this will be changed to a simpler API, similar to https://golang.org/pkg/os/exec/

Instead of .set_args(), the path will be set in the  new_process() function for example.